lib/linked_list.rb

```diff
@@ -14,75 +14,171 @@ def initialize(value, next_node = nil)
+    # Space Complexity: O(n)
     def search(value)
-      raise NotImplementedError
+      return false if @head.nil?
+      current = @head
+
+      until current.nil?
+        if current.data == value
+          return true
+        end
+        current = current.next
+      end
+
+      return false
     end
 
     # method to return the max value in the linked list
     # returns the data value and not the node
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def find_max
-      raise NotImplementedError
+      return nil if @head.nil?
+      current = @head
+      max_value = @head.data
+
+      until current.nil?
+        if current.data > max_value
+          max_value = current.data
+        end
+        current = current.next
+      end
+
+      return max_value
     end
 
     # method to return the min value in the linked list
     # returns the data value and not the node
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def find_min
-      raise NotImplementedError
+      return nil if @head.nil?
+      current = @head
+      min_value = @head.data
+
+      until current.nil?
+        if current.data < min_value
+          min_value = current.data
+        end
+        current = current.next
+      end
+
+      return min_value
     end
 
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def length
-      raise NotImplementedError
+      return 0 if @head.nil?
+      current = @head
+      count = 0
+
+      until current.nil?
+        count += 1
+        current = current.next
+      end
+
+      return count
     end
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns nil if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def get_at_index(index)
-      raise NotImplementedError
+      return nil if @head.nil?
+      count = 0
+      current = @head
+
+      until current.nil?
+        if count == index
+          return current.data
+        end
+        current = current.next
+        count += 1
+      end 
+
+      return nil
     end
 
     # method to print all the values in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def visit
-      raise NotImplementedError
+      return nil if @head.nil?
+      current = @head
+
+      until current.nil?
+        print current.data
+        current = current.next
+      end
     end
 
     # method to delete the first node found with specified value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def delete(value)

```

Comment:
👍 , The space complexity is O(1) because you're not making a new list with this method.